liquidity_pool/src/utils.rs

```diff
@@ -134,13 +135,69 @@ pub trait UtilsModule:
         )
     }
 
+    fn update_borrow_index(&self, round_last_update: u64) {
+        let borrow_rate = self.get_borrow_rate();
+        let delta_rounds = self.get_round_diff(round_last_update);
+
+        self.borrow_index()
+            .set(self.borrow_index().get() + &borrow_rate * delta_rounds);
+    }
+
+    fn update_supply_index(&self, rewards_increase: BigUint) {
+        let reserve_amount = self.reserves().get();
+        let borrowed_amount = self.borrowed_amount().get();
+        let rewards_reserves_paid = self.rewards_reserves_paid().get();
+        let total_amount = &reserve_amount + &borrowed_amount - rewards_reserves_paid;
```

Comment:
I assume that `rewards_reserves_paid` is subtracted here because supply interest should not be compounded (though see my previous comment about `rewards_reserves_paid` - I don't fully understand this value). However, that means that users gain more if they periodically withdraw their investment, then they deposit it back together with the gained interest. Is that correct? Is that desirable?

Assuming it is, this strategy is not available to borrowers, whose `lend_tokens` are held as collateral. Is it reasonable that lenders that are not also borrowers can gain more interest than lenders that are borrowers?

Reply:
I added more detail in this issue which we use for tracking stuff: https://github.com/runtimeverification/elrond-internal/issues/181

Reply:
Indeed, the lenders-not-borrowers can manually have compounded interest by withdrawing and lending again the amount with the interest. This is the case for farming as well.

This is a desirable behavior. We want to favor the lenders-not-borrowers because they provide liquidity for the pool.
How does this decreases the lender-and-borrower gain? We will have to pay the `debt for borrow` - `interest for supply`.

Reply:
Below I will try to explain what I meant by "decreases the lender-and-borrower gain". At the same time, I think that in some cases lenders' individual incentives will make them decrease their own profits. This may be fine, but I'd like you to take a look at it, so I'm including it here.

The summary
==========

* Borrowers receive less interest if the other lenders reinvest their interest because the incoming interest is distributed to more "shares" (but that's not the only reason)
* Each lender that is not also a borrower has an incentive to reinvest their interest often. However, in aggregate, that reduces the utilization rate, which reduces the incoming interest, which may or may not reduce the interest received by lenders-not-borrowers. To be specific, their interest is decreased by the reduction in utilization rate, but is increased by having more "shares" in the pool (this increase is at the expense of lenders that are also borrowers), and it depends which is larger.

Detailed computation
================

Changes in interest when the deposited value changes
-----------------------------------------------------------------------
For a pool, let us assume that
* The deposited amount is `D` (does not include interest, part of it may be currently borrowed)
* The borrowed amount is `Bp`
* The pool is underutilized (makes the formulas simpler)
* Someone adds (deposits) `x` tokens to the pool

Here is how the interest changes:

The interest `I` is computed as `Bp*brate=Bp*Up*alpha=Bp*(Bp/D)*alpha=Bp*Bp*α/D` where `Up` is the utilization and `α` is a constant.

Then we can compute the interest before adding the `x` tokens as `I = Bp*Bp*α/D`, and the interest after adding them as `Ix = Bp*Bp*α/(D+x)`.

The interest per invested token can be computed as `i = Bp*Bp*α/D/D = (Bp/D)^2 * α` and `ix = (Bp/(D+x))^2 * α`

Changes in interest for specific users
------------------------------------------------

Let us look at what happens to two users, L (lender) and B (Borrower). There are two pools, P and Q.

L deposited `ld` tokens in P and B deposited `bd` tokens in P. P has `D` deposited tokens in total, including `ld and `bd`. There are `Bp` tokens borrowed from `P`.

B also borrowed some amount from Q using `bd` as colalteral, but the actual amount is not important for this example, the important thing is that B's lending tokens are used as collateral in Q.

Let's also assume that L has accumulated `li` interest in P. Interest accumulated by B is not important.

* If L does nothing, then, in the next time period, L gets `ld*(Bp/D)^2 * α` and B gets `bd*(Bp/D)^2 * α` as interest.

* If L removes his tokens and reinvests, L has a deposit of `ld+li`. L gets `(ld+li)*(Bp/(D+li))^2 * α` as interest, while B gets `bd*(Bp/(D+li))^2 * α`.

*Note that B's gain is lower in the the second case.*

To finish the example, for many plausible values of `ld` and `li`, L gets more in the second case than in the first case (I think that the requirement is that `ld*(ld+li) < (D-ld)^2`).

I other words, by reinvesting, up to some level, lenders can increase their share of the pool, which decreases the share of anyone not reinvesting. It could be argued that lenders do not add additional liquidity in this case, since the system seems to use accumulated interest for borrowing, however, IIRC, reinvesting would decrease the pool utilization, so maybe this reinvesting does something.

As you mentioned above, this may be ok since you want to favor lenders-not-borrowers.

Lenders following their incentive
----------------------------------------

I have made the following two points above:

1. Lenders for which `ld*(ld+li) < (D-ld)^2` have an incentive to reinvest their interest, since they would earn more.
2. Assuming that there is at least one investor taking advantage of (1), everyone who does not reinvest their interest receives a lower interest than if (1) happened.

All users for which the equation at item (1) is true would have an incentive to reinvest often, which means that `li` is very small, so the relation is approximated by `ld^2 < (D-ld)^2`, i.e. `2*ld < D`. My guess is that this would usually be true almost always, but who knows. Anyway, people would have less incentive to to have their total investment be more than half of the amount in P, which provides some evidence for my guess.

Let us assume that all users have an incentive to reinvest, and let us see what happens if multiple lenders reinvest their interest. Let us assume that the total interest reinvested is `LI`, and let us see what happens to the same user L:

As above, without any reinvesting, L would get `ld*(Bp/D)^2*α`. With everyone reinvesting, L gets `(ld+li)*(Bp/(D+LI))^2*α`. I think that L gets more tokens than in the first case if `D/LI >= (ld/li)*(1 + sqrt(1 + li/ld))`.

To see why that's the case, note that `ld*(Bp/D)^2*α <= (Bp/(D+LI))^2*α` is equivalent to `0 <= li*D^2 - 2*D*LI*ld - ld*LI^2`, and the expression on the right is a second degree polynomial in `D`, which is positive if `D` is not between the polynomial's roots, i.e. if `D` is not between `(ld +- sqrt(ld(ld+li))) * LI/li`. However, `D` is positive and `(ld - sqrt(ld(ld+li))) * LI/li` is negative, so we only need `D >= (ld + sqrt(ld(ld+li))) * LI/li`, which can be rewritten as `D/LI >= (ld/li)*(1 + sqrt(1 + li/ld))`.

As above, most likely everyone reinvests often, so `li` is small, and li/ld is very small. This means that we can approximate `1 + li/ld` with `1`, so we get `D/LI >= (ld/li)*2`, or `D/ld >= 2*LI/li`.

D is composed by the amount deposited by lenders-not-borrowers (LD) and by lenders-borrowers (BD), i.e. `D = LD+BD`. We can approximate `LI/li` with `LD/ld`, so the relation above becomes `(LD+BD)/ld >= 2*LD/ld`, i.e. `BD/ld >= LD/ld`, i.e. `BD >= LD`.

If this is correct, it means that
1. Lenders are incentivized to reinvest their interest since each individual reinvestment is better than non-reinvestment.
2. Overall, lenders gain more with reinvesting than without only if `BD >= LD`, i.e. if the amount invested in the P pool by lenders-borrowers is larger than the amount invested by lenders-not-borrowers. If `BD < LD`, lenders gain less when all of them are reinvesting.

Does this seem reasonable?

Reply:
Wow, really great analysis @virgil-serbanuta ! Thank you!
Yes, it seems very reasonable the assumptions you made and result you came to.

As a design, we want to favor users to add more liquidity in the contract, so reinvesting is a really good choice. We want the interest accumulated (`LI`) to be reinvested instead of sitting in the contract without adding any benefit.

Furthermore, more liquidity in the contract means more funds available for new borrowers and, a lower `borrow rate` for the current borrowers which, in the end, means cheaper borrowing fees. Therefore, borrowers are favored by lenders-not-borrowers (`LD`) because they lower the fees by reinvesting their interest and lowering the capital utilization hence the borrow rate.

Thank you again for the analysis! Really great job!